src/Domain/InterfaceHelpers.hpp

```diff
@@ -76,29 +77,152 @@ struct unmap_interface_args<false> {
+          extra_args...);
+    }
+  }
+};
+
+template <typename InterfaceReturnType, typename DirectionsTag,
+          typename VolumeTagsList, typename... ArgumentTags>
+struct DispatchInterfaceInvokable<false, InterfaceReturnType, DirectionsTag,
+                                  VolumeTagsList, ArgumentTags...> {
+  template <typename InterfaceInvokable, typename DbTagsList,
+            typename... ExtraArgs>
+  static constexpr auto apply(InterfaceInvokable&& interface_invokable,
+                              const db::DataBox<DbTagsList>& box,
+                              ExtraArgs&&... extra_args) noexcept {
+    DirectionMap<DirectionsTag::volume_dim, InterfaceReturnType> result{};
+    for (const auto& direction : get<DirectionsTag>(box)) {
+      auto interface_value = interface_invokable(
+          unmap_interface_args<
+              tmpl::list_contains_v<VolumeTagsList, ArgumentTags>>::
+              apply(direction,
+                    get<tmpl::type_from<make_interface_tag_impl<
+                        ArgumentTags, DirectionsTag, VolumeTagsList>>>(box))...,
+          extra_args...);
+      result.insert({direction, std::move(interface_value)});
+    }
+    return result;
+  }
+};
+
+template <typename DirectionsTag, typename VolumeTagsList,
+          typename... ArgumentTags>
+struct DispatchInterfaceInvokable<false, void, DirectionsTag, VolumeTagsList,
+                                  ArgumentTags...> {
+  template <typename InterfaceInvokable, typename DbTagsList,
+            typename... ExtraArgs>
+  static constexpr void apply(InterfaceInvokable&& interface_invokable,
+                              const db::DataBox<DbTagsList>& box,
+                              ExtraArgs&&... extra_args) noexcept {
+    for (const auto& direction : get<DirectionsTag>(box)) {
+      interface_invokable(
+          unmap_interface_args<
+              tmpl::list_contains_v<VolumeTagsList, ArgumentTags>>::
+              apply(direction,
+                    get<tmpl::type_from<make_interface_tag_impl<
+                        ArgumentTags, DirectionsTag, VolumeTagsList>>>(box))...,
+          extra_args...);
+    }
+  }
+};
+
+template <typename DirectionsTag, typename ArgumentTags, typename VolumeTags>
+struct InterfaceApplyImpl;
+
+template <typename DirectionsTag, typename VolumeTagsList,
+          typename... ArgumentTags>
+struct InterfaceApplyImpl<DirectionsTag, tmpl::list<ArgumentTags...>,
+                          VolumeTagsList> {
+  static constexpr size_t volume_dim = DirectionsTag::volume_dim;
+
+  template <typename InterfaceInvokable, typename DbTagsList,
+            typename... ExtraArgs,
+            Requires<is_apply_callable_v<
+                InterfaceInvokable,
+                const db::const_item_type<ArgumentTags, DbTagsList>&...,
+                ExtraArgs...>> = nullptr>
+  static constexpr auto apply(InterfaceInvokable&& interface_invokable,
```

Comment:
why auto when you know the return type?

Reply:
It's not trivial to get the return type at this point. I could convert the alias in line 196 into a default template parameter to this function and then write another metafunction that maps the `interface_return_type` to `DirectionMap<Dim, interface_return_type>` unless it is `void`. It seemed easier to me to have the compiler infer the return type with `auto` instead. What do you think?